### ISSUE TITLE
V.prototype.text creates text nodes with <tspan x="0" /> after upgrade to joint 1.10

### ISSUE THREAD

**Issue by alp82** [clientIO/joint#566]
The problematic line in `V.prototype.text` is:

    var x = this.attr('x') || 0;

In my case that causes all custom elements to reset the x coordinate in all text elements. `<text>` now contains `<tspan>` elements that have the attribute `x="0"`. That overrides the x values of the parent text node and appears to render all my text left-aligned.

This can be fixed by simply replacing that line with:

    var x = this.attr('x');

Tested on Windows 10, Chrome 57

**Comment by kumilingus** [comment]
Hi, are you sure this was not an issue in `v1.0.3`? Looking at the changes the `x` always defaulted to `0`. Cheers.

**Comment by alp82** [comment]
Yes i am sure that issue was new when upgrading from 1.0.3 to 1.1.0. Maybe something else caused this? Where was that `x` used in 1.0.3?

**Comment by kumilingus** [comment]
https://github.com/clientIO/joint/blob/v1.0.3/src/vectorizer.js#L329
https://github.com/clientIO/joint/blob/v1.0.3/src/joint.dia.element.js#L465-L468

Would you mind posting here the custom element definition of yours here, so I can investigate further?

**Comment by alp82** [comment]
I will, but cannot do until Monday.

**Comment by luis-carlos-campos** [comment]
I've got the same problem here. +1

**Comment by alp82** [comment]
Woah, i am sorry for not responding. I completely lost track of this issue.

Our base node class is very large, with many custom methods. I'll only put the (hopefully) relevant bits in here that are related to rendering the node.

The model:
```
joint.shapes.cam.Control = joint.shapes.devs.Model.extend({
    defaults: joint.util.deepSupplement({
        type: "cam.Control",
        markup: [
            '<g class="rotatable">' +
            '<rect class="body" />' +
            '<rect class="header-body" />' +
            '<g class="header">' +
            '<image class="icon" />' +
            '<text class="label" />' +
            '</g>' +
            '<g class="portItems" />' +
            '<rect class="drop-target" />' +
            '<text class="edit-button" />' +
            '</g>'
        ].join(),
        wrapper: null,
        controlType: null, /* name of the control type, i.e. foreach, get, parser */
        configItemExports: [], /*  */
        config: {}, /* currently set config on the control */
        portConfigs: {},
        label: null, /* label of the control */
        icon: "resource/pem/material/ic_settings_white.svg", /* icon next to the label */
        primaryColor: "#586F7C", /* main color, used as background */
        secondaryColor: "#ffffff", /* support color, use as text color */
        orderIndex: 600,
        editor: null, /* classname of editor which will be instantiated on demand and allows editing the value of the control */
        description: "A visual editor control. Right click for options.",
        size: {width: 150, height: 32},
        inPortProperties: null,
        outPortProperties: null,
        hasCustomScript: false,
        outPortContexts: {},
        portPropertyDescriptors: null,
        attrs: {
            '.': {
                cursor: 'pointer'
            },
            '.body': {
                fill: 'white',
                stroke: 'grey'
            },
            '.header-body': {
                ref: '.body',
                'ref-width': 1,
                'ref-height': 1
            },
            '.icon': {
                x: 6,
                y: 4,
                width: 24,
                height: 24
            },
            '.label': {
                x: 40,
                'ref-x': null,
                'ref-y': null,
                y: 20,
                'y-alignment': null,
                'font-size': 12,
                'text-anchor': null
            },
            '.drop-target': {
                'port-group': 'drop-target',
                display: 'none',
                'ref': '.body',
                'ref-x': 0.5,
                'ref-width': -2,
                'x-alignment': 'middle',
                height: 30,
                y: 30,
                fill: 'orange',
                cursor: 'pointer'
            },
            '.edit-button': {
                text: '[ edit ]',
                display: 'none',
                'ref': '.body',
                'ref-x': 0.5,
                y: 45,
                'x-alignment': 'middle',
                height: 20,
                'font-size': 10,
                fill: '#444444',
                cursor: 'pointer'
            }
        },
        ports: {
            groups: {
                'in': {
                    attrs: {
                        '.port-body': {
                            magnet: 'passive',
                            fill: 'white',
                            stroke: 'grey',
                            r: 5
                        },
                        '.primary-key.port-body': {
                            fill: 'lightgoldenrodyellow',
                            r: 0
                        },
                        '.port-label': {
                            opacity: 0,
                            'font-size': 10
                        }
                    }
                },
                'out': {
                    attrs: {
                        '.port-body': {
                            fill: 'white',
                            stroke: 'grey',
                            r: 5
                        },
                        '.port-label': {
                            opacity: 0,
                            'font-size': 10
                        }
                    }
                }
            }
        }
    }, joint.shapes.devs.Model.prototype.defaults),
    
    [...]
});
```

and the corresponding ElementView:

```
joint.shapes.cam.ControlView = joint.dia.ElementView.extend({

    events: {
        "mouseover": "onpointerover",
        "mouseout": "onpointerout"
    },

    initialize: function () {
        this.__highlighted = {};
        joint.dia.ElementView.prototype.initialize.apply(this, arguments);

        this.listenTo(this.model, 'change:config', this._recalculateSize, this);
        this._recalculateSize();

        this.listenTo(this.model, "change:validation", this._onValidationChanged, this);
        this._onValidationChanged();
    },

    onpointerover: function (e) {
        this.highlight();
        var validation = this.model.get('validation') || {};
        if (this.model.get('isMenuMode') || this.model.get('enableTooltip') || validation.severity) {
            this.showTooltip(e);
        }
    },

    onpointerout: function () {
        this.unhighlight();
        this.hideTooltip();
    },


    _onLabelChanged: function () {
        var label = this.model.getLabel();
        this.model.attr({
            '.label': {
                text: label
            }
        });
    },

    _onIconChanged: function () {
        var icon = this.model.getIcon();
        this.model.attr({
            '.icon': {
                'xlink:href': icon
            }
        });
    },

    _onPrimaryColorChanged: function () {
        var primaryColor = this.model.getPrimaryColor();
        this.model.attr({
            '.header-body': {
                fill: primaryColor
            }
        });
    },

    _onSecondaryColorChanged: function () {
        var secondaryColor = this.model.getSecondaryColor();
        this.model.attr({
            '.label': {
                fill: secondaryColor
            }
        });
    },

    _onDescriptionChanged: function () {
        if (this._tt) {
            this._tt.innerHTML = this.model.get('description') || '';
        }
    },

    _recalculateSize: function () {
        var newSize = this._calcSize();
        this.model.set("size", newSize);
    },

    _onValidationChanged: function () {
//        var validation = this.model.get('validation') || {};
        this._onPrimaryColorChanged();
    },

    _calcSize: function () {
        var serverModel = this.model.get("serverModel");
        var inportcount = (this.model.get("inPorts") || []).length;
        var outportcount = (this.model.get("outPorts") || []).length;
        var offsetTop = this._getOffsetTop();
        var offsetBottom = this._getOffsetBottom();

        var maxPortCount = Math.max(1, inportcount, outportcount);
        var height = offsetTop + maxPortCount * 20 + offsetBottom;

        var width;
        if (this.model.get('isMenuMode')) {
            width = Math.max(110, (this.model.get("size") || {}).width);
        } else {
            var label = this.model.get("label") || "";
            width = Math.max(110, label.length * 7 + 55);
        }

        var size = {
            width: width,
            height: height
        };
        return size;
    },

    _getOffsetTop: function () {
        return 5;
    },

    _getOffsetBottom: function () {
        return 5;
    },

    showTooltip: function (e) {
        var description = this.model.get('description');
        var validation = this.model.get('validation') || {};
        if (description || validation.severity) {
            if (!this._tt) {
                /* body mouse listener for placing the dragged element */
                var bodyListener = function (e) {
                    this.__ttPosX = e.clientX;
                    this.__ttPosY = e.clientY;
                }.bind(this);
                document.body.addEventListener("mousemove", bodyListener);

                var container = this._tt = document.createElement("div");
                container.setAttribute('style', 'position:absolute; top: ' + (e.clientY + 15) + 'px; left: ' + (e.clientX + 15) + 'px; ' +
                    'max-width: 200px; background: #fafafa; color: black; z-index: 9999999; font-family: Tahoma, Verdana, sans-serif; ' +
                    'padding: 5px; font-size: 12px; border: 1px solid #ddd; display: none;');
                document.body.appendChild(container);
            }
            this._tt.innerHTML = validation.severity ? validation.severity + ': ' + (validation.content || validation.label) : description;
            this.__ttDelayId = setTimeout(function () {
                this._tt.style.display = 'block';
                window.requestAnimationFrame(this.__animateTooltip.bind(this));
            }.bind(this), 200);
        }
    },

    hideTooltip: function () {
        clearTimeout(this.__ttDelayId);
        if (this._tt) {
            this._tt.style.display = 'none';
        }
    },

    __animateTooltip: function () {
        if (this.__ttPosX == null) {
            return;
        }
        this._tt.style.top = (this.__ttPosY + 15) + "px";
        this._tt.style.left = (this.__ttPosX + 15) + "px";
        if (this._tt.style.display !== 'none') {
            window.requestAnimationFrame(this.__animateTooltip.bind(this));
        }
    },

    highlight: function () {
        var selector = typeof arguments[0] === "string" ? arguments[0] : '.body';
        var elem = typeof arguments[0] === "object" ? arguments[0] : this.el.querySelector(selector);
        if (elem) {
            if (this.__highlighted[elem.id]) {
                return;
            }
            this.__highlighted[elem.id] = {
                stroke: elem.getAttribute("stroke") || "black",
                strokeWidth: elem.getAttribute("stroke-width") || 1
            };
            elem.setAttribute("stroke", "orange");
            elem.setAttribute("stroke-width", 3);
        }
    },

    unhighlight: function () {
        var selector = typeof arguments[0] === "string" ? arguments[0] : '.body';
        var elem = typeof arguments[0] === "object" ? arguments[0] : this.el.querySelector(selector);
        if (elem) {
            var originalData = this.__highlighted[elem.id];
            if (!originalData) {
                /* not yet highlighted */
                return;
            }
            elem.setAttribute("stroke", originalData.stroke);
            elem.setAttribute("stroke-width", originalData.strokeWidth);
            this.__highlighted[elem.id] = null;
        }
    },

    render: function () {
        var serverModel = this.model.get("serverModel");
        if (this.model.get('isMenuMode')) {
            var title = this.el.querySelector("title");
            if (title) {
                title.textContent = "Doubleclick to add.";
            }
            this.model.set("inPorts", []);
            this.model.set("outPorts", []);
        }

        joint.dia.ElementView.prototype.render.apply(this, arguments);
    }
});

```

**Comment by luis-carlos-campos** [comment]
@kumilingus Have you had a chance to take a look at it?

**Comment by luis-carlos-campos** [comment]
I don't think its the proper solution, but if you have a multiple lines text and you need to make it work, you can do something like that:

        var x = this.attr('x');

        if (lines.length > 1) {
            x = x || 0;
        }

Cheers,

**Comment by alp82** [comment]
Any news here? Would love to revert my code changes to the joint.js sources.

**Comment by kumilingus** [comment]
Hi @alp82, here is a workaround for the time being. I am sorry I am way too busy this week.
```javascript
'.label': {
                transform: 'translate(40,0)',  // instead of `x: 40`,
                y: 20,
                'ref-x': null,
                'ref-y': null,
                'y-alignment': null,
                'font-size': 12,
                'text-anchor': null
            },
```

**Comment by kumilingus** [comment]
Also what is proposed here by @luis-carlos-campos looks good to me (at least it works well on Chrome). Would you mind creating a PR?

**Comment by kumilingus** [comment]
Fixed in #625.

**Comment by luis-carlos-campos** [comment]
@kumilingus Sorry about the PR, I was travelling. Thank you so much for the fix.

Regards,
